Distribution/Server/Features/ReverseDependencies/State.hs

```diff
@@ -13,406 +15,203 @@ import Distribution.PackageDescription
+instance MemSize ReverseIndex where
+    memSize (ReverseIndex a b c) = memSize3 a b c
 
-    -- Cached counts for reverse dependencies.
-    reverseCount :: Map PackageName ReverseCount
-} deriving (Eq, Show, Typeable)
 
 emptyReverseIndex :: ReverseIndex
-emptyReverseIndex = ReverseIndex (PackageIndex.fromList [])
-                                 Map.empty Map.empty Map.empty
-
-data ReverseCount = ReverseCount {
-    directReverseCount :: Int,
-    flattenedReverseCount :: Int,
-    versionReverseCount :: Map Version Int
-} deriving (Show, Eq, Typeable, Ord)
-emptyReverseCount :: ReverseCount
-emptyReverseCount = ReverseCount 0 0 Map.empty
+emptyReverseIndex = ReverseIndex (PackageIndex.fromList []) (Gr.buildG (0, 20000) []) Bimap.empty
 
 constructReverseIndex :: PackageIndex PkgInfo -> ReverseIndex
 constructReverseIndex index =
-    let deps = constructRevDeps index
-    in updateReverseCount $ emptyReverseIndex {
+    emptyReverseIndex {
         duplicatedIndex = constructDupIndex index,
-        reverseDependencies = deps,
-        flattenedReverse = packageNameClosure deps
-    }
-
-constructDupIndex :: PackageIndex PkgInfo -> PackageIndex PackageId
-constructDupIndex = PackageIndex.fromList
-  . map packageId
-  . PackageIndex.allPackages
-
-updateReverseCount :: ReverseIndex -> ReverseIndex
-updateReverseCount index =
-    let deps = reverseDependencies index
-        flat = flattenedReverse index
-    in index {
-        reverseCount = flip Map.mapWithKey deps $ \pkg (versions, _) -> ReverseCount {
-            directReverseCount = Map.size . Map.unions $ Map.elems versions,
-            flattenedReverseCount = maybe 0 Set.size $ Map.lookup pkg flat,
-            versionReverseCount = Map.map Map.size versions
-        }
+        reverseDependencies = constructRevDeps index nodePkgMap,
+        packageNodeIdMap = nodePkgMap
     }
-
-addPackage :: PackageId -> CombinedDeps
-           -> ReverseIndex -> (ReverseIndex, Map PackageName [Version])
-addPackage pkgid deps revs =
-    let index = PackageIndex.insert pkgid (duplicatedIndex revs)
-        (rd, rec) = registerPackage (getAllVersions index) pkgid deps (reverseDependencies revs)
-    in (updateReverseCount $ revs {
+    where
+        nodePkgMap = foldr (uncurry Bimap.insert) Bimap.empty $ zip (map packageId (PackageIndex.allPackages index)) [0..]
+        constructDupIndex :: PackageIndex PkgInfo -> PackageIndex PackageId
+        constructDupIndex = PackageIndex.fromList
+          . map packageId
+          . PackageIndex.allPackages
+
+addPackage :: PackageId -> [PackageId]
+           -> ReverseIndex -> ReverseIndex
+addPackage pkgid deps ri@(ReverseIndex pindex revs nodemap) =
+    let index = PackageIndex.insert pkgid pindex
+        rd' | revSize revs < Bimap.size nodemap = arrDouble revs | otherwise = revs
+        npm = Bimap.tryInsert pkgid (Bimap.size nodemap) nodemap
+        rd = insEdges (map (\d -> (npm ! d, [npm ! pkgid])) deps) rd'
```

Comment:
This graph enlarging ought to be done differently. Array size doubling is bogus for pure immutable arrays.